### DISCUSSION TITLE
Customizing Wix

### DISCUSSION THREAD

**Discussion by hrzafer** [wiverson/jtoolprovider-plugin#10] [Q&A]
Hi, 

I am trying to customize my Windows installer. [Here](https://docs.oracle.com/en/java/javase/16/jpackage/override-jpackage-resources.html#GUID-1B718F8B-B68D-4D46-B1DB-003D7729AAB6) it says, we can use `--resource-dir` to tell `jpackage` to override some of the Wix configuration.  I tried passing that option into the `win-packages.txt` file as 
```
--resource-dir  ${project.build.directory}\packaging
```

and creating a `main.wxs` file under `src/packaging`. However, it still uses the `main.wxs` created under `target/installer-work/config`.  Does this plugin directly or indirectly support this kind of customization?

Thanks.

**Comment by wiverson** [comment]
The plugin just calls jpackage via a JDK supplied API (as opposed to, say, kicking it off via the command line).

I haven't messed around with actually using that customization. From the jpackage docs looks like --verbose and --temp are helpful for debugging. It kind of looks like the process is to use --temp to generate the default file directory structure, then drop in override files, and then run jpackage again...?

The plugin should be perfectly happy running jpackage command. Depending on your Maven expertise, you might want to try:

- running jpackage with --temp to get a working file set
- drop an updated main.wxs into that directory
- run jpackage a second time with --resource-dir pointing at the temp directory

You could absolutely do all that via the plugin and a resources plugin execution. How is your Maven? ;)

**Reply by hrzafer** [reply]
I know maven but I don't know much about wix. It is hard to customize the installation process.

**Reply by wiverson** [reply]
I'm basically in a similar spot. I bought a book on Wix and it's yet another hilariously specialized bit of tech.

My two cents - basically, Wix lets you build and work with Microsoft MSI files. MSI files are basically glorified little databases, with a million little special flags and tweaks for different scenarios. This file shows how an MSI database can be manipulated via JavaScript:

https://github.com/wiverson/maven-jpackage-template/blob/main/src/packaging/add-launch-to-msi.js

This particular file is run to tweak the installer to launch the built executable on launch. I got the guts of this by posting on a Windows installer subreddit.

A lot of the resources and tools for Windows installers date back to the early 2000s, and it shows in terms of the tech, documentation, etc.

Basically, jpackage generates Wix file[s] and then runs the Wix tooling to generate the msi. In this template I use that JavaScript to tweak the generated msi. Depending on your needs, you might want to either tweak the msi or the generated Wix file, which is dropped into a temp/working directory if the right flags are passed in to jpackage.

Good luck. You'll feel like you are rocking the early 2000s, so maybe dig up some music from that period to help set the mood. :)
